### PR TITLE
awskinesisexporter: Refactor AWS kinesis exporter to be synchronous 

### LINKED ISSUE #317: Add README for kinesis exporter
https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/exporter/kinesisexporter is missing a README.md.

### LINKED ISSUE #1305: [kinesisexporter] Support publishing of multiple data types in the kinesis exporter.
**Is your feature request related to a problem? Please describe.** 
Leverage the kinesisexporter to export otlp protobuff to kinesis streams. Currently, the format is fixed to the jaegar/zipkin format. In a more general sense, we would like support for format-agnostic, where specific data formats can be published by implementing marshaller classes.

**Describe the solution you'd like**
Use the marshallers pattern that the core collector library uses within the [kafka exporter.](https://github.com/open-telemetry/opentelemetry-collector/blob/master/exporter/kafkaexporter/marshaller.go) An otlp marshaller would be added to support the otlp protobuff format. A jaeger marshaller could also be added to support existing capabilities.

This requires swapping out the currently used producer library which only supports the jaeger format to a format-agnostic one. In our current fork we have chosen the [signalfx omnition producer](https://github.com/signalfx/omnition-kinesis-producer).

**Describe alternatives you've considered**
- Writing a new kinesis exporter - duplicate work as we could leverage existing code.

**Additional context**
- We have already implemented a functional version on the [atlassian fork](https://github.com/atlassian-forks/opentelemetry-collector-contrib/pull/9/files) of the repository. Some standardisation and feedback is needed to get it up to standard for a PR. 

- We've also added support for metrics marshalling and unit tests for files which lacked coverage. [PR link](https://github.com/atlassian-forks/opentelemetry-collector-contrib/pull/9/files)

### PR DESCRIPTION
**Pull request by MovieStoreGuy**
**Description:**
I have update the kinesis exporter to mirror the behaviour seen within the Kafka Exporter in order to benefit the existing helper functions and keep exporting to streaming services consistent.

This changes the existing behaviour that would send data in an async, this is problematic since once the data was accepted by the exporter, any issues trying to publish to the stream was lost and hard to reconcile how much data was dropped as a result of this. 

**Link to tracking Issue:** Partially resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1305
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/317

**Testing:**
All added components have new unit tests and mock integration testing where sending to kinesis is involved.

**Documentation:** 
The readme for the exporter has been added explaining the intent of the exporter and alluding to remaining work which is to allow other encoding types beyond jaeger.

### CODE REVIEW

File: cmd/configschema/go.mod [hunk 1]
```diff
@@ -44,7 +44,7 @@ require (
 	github.com/alecthomas/units v0.0.0-20210208195552-ff826a37aa15 // indirect
```
Comment (MovieStoreGuy) [inline cmd/configschema/go.mod:44]: These changes were a result of me trying to get the linting to pass.

**Comment by MovieStoreGuy** [comment]
Sorry @owais , I had promised I get this done some time ago. It unfortunately slipped off my list of things to do.

File: exporter/awskinesisexporter/internal/batch/batch.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+// to be written to he kinesis.PutRecords endpoint
+func (b *Batch) Chunk() [][]*kinesis.PutRecordsRequestEntry {
+	length, chunk := len(b.records), make([][]*kinesis.PutRecordsRequestEntry, 0, len(b.records)/b.maxBatchSize+1)
+	for i, end := 0, min(length, b.maxBatchSize); i < length; i, end = i+b.maxBatchSize, i+end+min(b.maxBatchSize, length-end) {
+		chunk = append(chunk, b.records[i:end])
+	}
+	return chunk
+}
+
+func min(a, b int) int {
```
Comment (MovieStoreGuy) [inline exporter/awskinesisexporter/internal/batch/batch.go:125]: I can't tell you how much i hate having this here and eagerly await the promise of generics in go 1.18

File: exporter/awskinesisexporter/internal/producer/interface.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+package producer
+
+import (
+	"context"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awskinesisexporter/internal/batch"
+)
+
+// Batcher abstracts the raw kinesis client to reduce complexity with delievering dynamic encoded data.
```
Comment (anuraaga) [inline exporter/awskinesisexporter/internal/producer/interface.go:23]:
```suggestion
// Batcher abstracts the raw kinesis client to reduce complexity with delivering dynamic encoded data.
```

File: exporter/awskinesisexporter/internal/batch/batch.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		return err
+	}
+	return b.addRaw(data, key)
+}
+
+// Chunk breaks up the iternal queue into blocks that can be used
+// to be written to he kinesis.PutRecords endpoint
+func (b *Batch) Chunk() [][]*kinesis.PutRecordsRequestEntry {
+	length, chunk := len(b.records), make([][]*kinesis.PutRecordsRequestEntry, 0, len(b.records)/b.maxBatchSize+1)
+	for i, end := 0, min(length, b.maxBatchSize); i < length; i, end = i+b.maxBatchSize, i+end+min(b.maxBatchSize, length-end) {
```
Comment (anuraaga) [inline exporter/awskinesisexporter/internal/batch/batch.go:119]: This line is pretty gnarly, can we simplify it?

This pattern looks pretty readable, and (I think? :P) does the same logic

https://freshman.tech/snippets/go/split-slice-into-chunks/

**Review by anuraaga — APPROVED** [review]
Mostly looks good, just a small point

File: exporter/awskinesisexporter/internal/batch/batch.go (hunk 4 above)
Comment (MovieStoreGuy) [reply exporter/awskinesisexporter/internal/batch/batch.go:119]: You're not wrong, I will expand it out to make it easier to digest :)

**Comment by MovieStoreGuy** [comment]
@anuraaga Is it possible to get this merged?

**Comment by anuraaga** [comment]
@MovieStoreGuy We need a maintainer like @bogdandrutu or @tigrannajaryan to merge

**Pull request merged**